### PR TITLE
Pin firrtl version to 1.6.0-M2

### PR DESCRIPTION
**Pull request by mwachs5**
Using -SNAPSHOT does not make sense, because that means breaking changes in firrtl can break our C/I. At the moment, an upstream change has broken our documentation generation flow.

### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [ ] Did you delete any extraneous printlns/debugging code?
- [ ] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [ ] Did you state the API impact?
- [ ] Did you specify the code generation impact?
- [ ] Did you request a desired merge strategy?
- [ ] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

 - bug fix           

#### Backend Code Generation Impact

yes, it does, but only because upstream broke its generation

#### Desired Merge Strategy

   - Squash: The PR will be squashed and merged (choose this if you have no preference. 

#### Release Notes

Pin firrtl version to 1.6.0-M2

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.4.x`, [small] API extension: `3.5.x`, API modification or big change: `3.6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

**Review by jackkoenig — CHANGES_REQUESTED** [review]
> Using -SNAPSHOT does not make sense, because that means breaking changes in firrtl can break our C/I. 

On the contrary, that is the point. It is a negative consequence of having closely coupled projects in different repositories, but it is intentional.

**Comment by mwachs5** [comment]
I'm confused. How is this helpful ...? We can always change this back to SNAPSHOT when that's what we want it to be. Right now, nothing will make it through chisel3 CI.

**Comment by mwachs5** [comment]
though this PR is failing C/I in the same way anyway so I'm confused what's going on here.

**Comment by jackkoenig** [comment]
> I'm confused. How is this helpful ...? We can always change this back to SNAPSHOT when that's what we want it to be. Right now, nothing will make it through chisel3 CI.

You noticed the breakage immediately and now we should fix it. It's a way of catching compatibility breaking changes immediately rather than letting them build up until you decide to bump. This is because they are such closely coupled projects, obviously this is not necessarily a development process that every project should follow.

**Comment by mwachs5** [comment]
ok, right, again we can immediately put this back when the issue is resolved. Are we OK blocking on both chisel and CIRCT fixes to do any chisel3 work until this is resolved?

File: build.sbt [hunk 1]
```diff
@@ -3,7 +3,7 @@
 enablePlugins(SiteScaladocPlugin)
 
 val defaultVersions = Map(
-  "firrtl" -> "edu.berkeley.cs" %% "firrtl" % "1.6-SNAPSHOT",
+  "firrtl" -> "edu.berkeley.cs" %% "firrtl" % "1.6.0-M2",
   "treadle" -> "edu.berkeley.cs" %% "treadle" % "1.6-SNAPSHOT"
```
Comment (mwachs5) [inline build.sbt:7]:
```suggestion
  "firrtl" -> "edu.berkeley.cs" %% "firrtl" % "1.6.0-M2",
  "treadle" -> "edu.berkeley.cs" %% "treadle" % "1.6.0-M2"
```

**Comment by jackkoenig** [comment]
It depends on how hard the breakage is to fix but my suspicion is that it's super easy so why not just fix it?

**Comment by mwachs5** [comment]
> is that it's super easy so why not just fix it?

I was told this needs a fix to llvm/CIRCT, so we need to wait for CIRCT to fix the issue then do a release then bump it here

**Comment by ekiwi** [comment]
> I was told this needs a fix to llvm/CIRCT, so we need to wait for CIRCT to fix the issue then do a release then bump it here

We can also do a small fix in firrtl.

I was initially going to only add the RUW type if it is not undefined, but @seldridge asked that I add it unconditionally: https://github.com/chipsalliance/firrtl/pull/2595#discussion_r1086797340

We can make a PR to only conditionally emit it which should fix most of the breakage since using RUW explicitly is rare.

**Comment by mwachs5** [comment]
> We can also do a small fix in firrtl.

I mean, the goal is to move forward towards correctness, so if the current SFC behavior is where we want it to be, let's leave it there and keep pushing forward towards correctness. If no one else is blocked by this CI being broken we can leave this, if we are we can figure out a way to use "the last working version" until we have more pieces that work together...

**Comment by ekiwi** [comment]
here you go: https://github.com/chipsalliance/firrtl/pull/2597

**Comment by ekiwi** [comment]
> I mean, the goal is to move forward towards correctness, so if the current SFC behavior is where we want it to be, let's leave it there and keep pushing forward towards correctness.

Either behavior is correct. Just one is always explicit and the other leaves out the default when possible.

**Comment by jackkoenig** [comment]
I see, I agree with @ekiwi's change because breaking Chisel with all existing releases of firtool and SFC is ill-advised unless absolutely necessary. In cases where it is necessary, then I agree with @mwachs5's proposal to temporarily pin the version to keep making progress in the meantime. https://github.com/chipsalliance/firrtl/pull/2591 would also help for such things 😇

**Comment by mwachs5** [comment]
closing since SFC fix should get -SNAPSHOT working well again

**Pull request closed without merging**